src/browser/commands/saveState/index.ts

```diff
@@ -0,0 +1,174 @@
+import fs from "fs-extra";
+
+import type { Browser } from "../../types";
+// import { DumpIndexDB } from "./dumpIndexedDB";
+import { dumpStorage, StorageData } from "./dumpStorage";
+import { DEVTOOLS_PROTOCOL, WEBDRIVER_PROTOCOL } from "../../../constants/config";
+import { Cookie } from "@testplane/wdio-protocols";
+
+export type SaveStateOptions = {
+    path?: string;
+
+    cookies?: boolean;
+    localStorage?: boolean;
+    sessionStorage?: boolean;
+    // indexDB?: boolean;
+};
+
+export type FrameData = StorageData & {
+    // indexDB?: Record<string, DumpIndexDB>;
+};
+
+export type SaveStateData = {
+    cookies?: Array<Cookie>;
+    framesData: Record<string, FrameData>;
+};
+
+export const defaultOptions = {
+    cookies: true,
+    localStorage: true,
+    sessionStorage: true,
+    // indexDB: false,
+};
+
+export const getWebdriverFrames = async (session: WebdriverIO.Browser): Promise<string[]> =>
+    session.execute<string[], []>(() =>
+        Array.from(document.getElementsByTagName("iframe"))
+            .map(el => el.getAttribute("src") as string)
+            .filter(src => src !== null && src !== "about:blank"),
+    );
+
+export default (browser: Browser): void => {
+    const { publicAPI: session } = browser;
+
+    session.addCommand("saveState", async (_options: SaveStateOptions = {}): Promise<SaveStateData> => {
+        const options = { ...defaultOptions, ..._options };
+
+        const data: SaveStateData = {
+            framesData: {},
+        };
+
+        switch (browser.config.automationProtocol) {
+            case WEBDRIVER_PROTOCOL: {
+                if (options.cookies) {
+                    const storageCookies = await session.storageGetCookies({});
+
+                    data.cookies = storageCookies.cookies.map(cookie => ({
+                        name: cookie.name,
+                        value: cookie.value.value,
+                        domain: cookie.domain,
+                        path: cookie.path,
+                        expires: cookie.expiry,
+                        httpOnly: cookie.httpOnly,
+                        secure: cookie.secure,
+                    }));
+                }
+
+                await session.switchToParentFrame();
```

Comment:
I don't understand why switch to parent frame here, then do it again on line 114

Moreover, I don't understand the logic here.
1. We get N iframes at the current level context.
2. We switch to parent frame N times (???)
3. We iterate from -1 to N (?)

And another issue is that we don't switch back to the frame user had active before. So calling saveState actually mutates browser state, while it should not.

Shouldn't we have about the following algorithm?
1. Save current frame id and start from top-level browser context
2. Recursively:
    1. Save storage values on current level
    2. Iterate over all frames at current level and do the same thing
3. Restore current frame id